keps/sig-node/4033-group-driver-detection-over-cri/README.md

```diff
@@ -0,0 +1,880 @@
+There currently are two different low-level management interfaces for cgroups:
+maninpulating the cgroupfs directly or using the systemd system management
+daemon to manage them. Currently, both the kubelet and the container runtime
+has a configuration setting for selecting the cgroup driver (cgroupfs or
+systemd). These settings must be in sync, both kubelet and the runtime
+configured to use the same driver as the two drivers are incompatible because
+of a different kind of cgroups hierarchy used in them. Having kubelet and the
+container runtime to use non-matching cgroup drivers can cause hard to
+understand failures in container creation or inconsistent resource allocation
+on the node. This – two independent configuration settings for the same thing –
+is a common cause for user errors. Instead of having a split brain situation,
+there should be a single source of truth for the cgroup driver.
+
+### Goals
+
+- make kubelet automatically use the same cgroup driver as the container
+  runtime, making it unnecessary to specify cgroup driver in kubelet
+  configuration
+- maintain backward compatibility to run kubelet on top of an older container
+  runtime that doesn’t support this new feature
+
+### Non-Goals
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a cluster administrator, I would like to simplify my node configuration by
+configuring fields just once, and not needing to synchronize between multiple
+processes.
+
+#### Story 2
+
+As a novice Kubernetes user, I would like to easily be able to start pods with
+all runtimes, even if they have differing opinions on defaults.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+Field adoption could be considered a risk, though the CRI implementations work
+closely with SIG-Node and the feature will move along with CRI implementation
+adoption.
+
+## Design Details
+
+### CRI API
+
+Extend the CRI protocol to inform the kubelet which cgroup driver should be
+used.
+
+```diff
+ message RuntimeStatus {
+     // List of current observed runtime conditions.
+     repeated RuntimeCondition conditions = 1;
++    // Configuration settings of the runtime
++    RuntimeConfiguration configuration = 2;
+ }
+ 
++message RuntimeConfiguration {
++    // Cgroup driver to use
++    CgroupDriver cgroup_driver = 1;
++}
++
++enum CgroupDriver {
++    CGGROUPFS = 0;
++    SYSTEMD = 1;
++}
+```
+
+The existing RuntimeStatus message (of the existing Status API endpoint) is
+used as this is being frequently queried by the kubelet, and is a place where
+the runtime tells the Kubelet about its state. The runtime will decide which
+CgroupDriver to choose based on existing methods: its own configuration.
```

Comment:
Does kubelet have any kinds of cgroup interaction before it gets response from `RuntimeStatus`?  possibly creation of cgroup for its own usage or QOS hierarchy?